lib/holepuncher.js

```diff
@@ -0,0 +1,288 @@
+const utp = require('utp-native')
+const os = require('os')
+const Timer = require('./timer')
+
+const BIRTHDAY_SOCKETS = 256
+const HOLEPUNCH = Buffer.from([0])
+const HOLEPUNCH_TTL = 5
+const DEFAULT_TTL = 64
+
+// const NAT_OPEN = Symbol.for('NAT_OPEN')
+const NAT_UNKNOWN = Symbol.for('NAT_UNKNOWN')
+const NAT_PORT_RANDOMIZED = Symbol.for('NAT_PORT_RANDOMIZED')
+
+module.exports = class Holepuncher {
+  constructor (address) {
+    const self = this
+
+    this.address = address
+    this.socket = null
+
+    this.localNetwork = null
+    this.remoteNetwork = null
+    this.localFriendly = false
+    this.remoteFriendly = false
+    this.lan = false
+
+    this._onmessage = onmessage
+    this._allSockets = []
+    this._holepunchTimer = null
+    this._lanTimer = null
+    this._stopped = false
+    this._isClient = false
+
+    // async states
+    this._openingSessions = null
+    this._holepunching = null
+    this._resolveConnection = null
+    this._onconnection = onconnection
+
+    this.connected = new Promise((resolve) => { this._resolveConnection = resolve })
```

Comment:
mentioned already, but could be good to find an alternative if not too problematic elsewhere